src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,447 @@
+            }
+
+            public ReadOnlySpan<char> Buffer { get; init; }
+
+            public void Dispose()
+            {
+                if (_backingArray != null)
+                {
+                    ArrayPool<char>.Shared.Return(_backingArray);
+                }
+            }
+        }
+
+        private static void ConvertUtf8ToUtf16(ReadOnlySpan<byte> utf8TypeName, out Utf16SharedBuffer utf16Buffer)
+        {
+            // Use quick conservative estimate for small strings
+            int needed = (utf8TypeName.Length < 1024)
+                ? Encoding.UTF8.GetMaxCharCount(utf8TypeName.Length)
+                : Encoding.UTF8.GetCharCount(utf8TypeName);
+
+            char[] buffer = ArrayPool<char>.Shared.Rent(needed);
+            int converted = Encoding.UTF8.GetChars(utf8TypeName, buffer);
+            utf16Buffer = new Utf16SharedBuffer(buffer, converted);
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe Interop.BOOL NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            try
+            {
+                string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+                TypeNameValue targetTypeName = new()
+                {
+                    Utf8TypeName = arg->Utf8String2,
+                    Utf8TypeNameLen = arg->StringLen2,
+                    FallbackAssembly = context->CurrentAssembly
+                };
+                context->ExternalTypeMap.Add(externalTypeName, targetTypeName);
+            }
+            catch (Exception ex)
+            {
+                context->CreationException = ExceptionDispatchInfo.Capture(ex);
+                return Interop.BOOL.FALSE; // Stop processing.
+            }
+
+            return Interop.BOOL.TRUE; // Continue processing.
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe Interop.BOOL NewProxyTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            Utf16SharedBuffer sourceTypeBuffer = new();
+            try
+            {
+                ConvertUtf8ToUtf16(new ReadOnlySpan<byte>(arg->Utf8String1, arg->StringLen1), out sourceTypeBuffer);
+                TypeName parsedSource = TypeNameParser.Parse(sourceTypeBuffer.Buffer, throwOnError: true)!;
+
+                // If the assembly name is not included in the type name, use the fallback assembly name.
+                string assemblyName = parsedSource.AssemblyName?.FullName ?? context->CurrentAssembly.FullName!;
```

Comment:
I do not think we need to materialize assemblyName as a string here. It should be enough to store Utf8String1 + fallback assembly as the key that we can type load lazily.

Reply:
Oh, you're right. This particular code logic was in place before I started down using `VersionResilientHashCode`. It was never updated after.